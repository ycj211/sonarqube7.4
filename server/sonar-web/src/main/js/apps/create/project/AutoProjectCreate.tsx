/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import RemoteRepositories from './RemoteRepositories';
import OrganizationInput from './OrganizationInput';
import IdentityProviderLink from '../../../components/ui/IdentityProviderLink';
import { AlmApplication, Organization } from '../../../app/types';
import {
  ORGANIZATION_IMPORT_BINDING_IN_PROGRESS_TIMESTAMP,
  ORGANIZATION_IMPORT_REDIRECT_TO_PROJECT_TIMESTAMP
} from '../organization/utils';
import { translate } from '../../../helpers/l10n';
import { save } from '../../../helpers/storage';

interface Props {
  almApplication: AlmApplication;
  boundOrganizations: Organization[];
  onProjectCreate: (projectKeys: string[], organization: string) => void;
  organization?: string;
}

interface State {
  selectedOrganization: string;
}

export default class AutoProjectCreate extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { selectedOrganization: this.getInitialSelectedOrganization(props) };
  }

  getInitialSelectedOrganization(props: Props) {
    if (props.organization) {
      return props.organization;
    } else if (props.boundOrganizations.length === 1) {
      return props.boundOrganizations[0].key;
    } else {
      return '';
    }
  }

  handleInstallAppClick = () => {
    save(ORGANIZATION_IMPORT_BINDING_IN_PROGRESS_TIMESTAMP, Date.now().toString(10));
    save(ORGANIZATION_IMPORT_REDIRECT_TO_PROJECT_TIMESTAMP, Date.now().toString(10));
  };

  handleOrganizationSelect = ({ key }: Organization) => {
    this.setState({ selectedOrganization: key });
  };

  render() {
    const { almApplication, boundOrganizations, onProjectCreate } = this.props;

    if (boundOrganizations.length === 0) {
      return (
        <>
          <p className="spacer-bottom">
            {translate('onboarding.create_project.install_app_description', almApplication.key)}
          </p>
          <IdentityProviderLink
            className="display-inline-block"
            identityProvider={almApplication}
            onClick={this.handleInstallAppClick}
            small={true}
            url={almApplication.installationUrl}>
            {translate(
              'onboarding.import_organization.choose_organization_button',
              almApplication.key
            )}
          </IdentityProviderLink>
        </>
      );
    }

    const { selectedOrganization } = this.state;
    return (
      <>
        <OrganizationInput
          autoImport={true}
          onChange={this.handleOrganizationSelect}
          organization={selectedOrganization}
          organizations={this.props.boundOrganizations}
        />
        {selectedOrganization && (
          <RemoteRepositories
            almApplication={almApplication}
            onProjectCreate={onProjectCreate}
            organization={selectedOrganization}
          />
        )}
      </>
    );
  }
}
