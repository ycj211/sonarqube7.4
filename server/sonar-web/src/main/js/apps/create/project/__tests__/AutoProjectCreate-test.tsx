/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import AutoProjectCreate from '../AutoProjectCreate';

const almApplication = {
  backgroundColor: 'blue',
  iconPath: 'icon/path',
  installationUrl: 'https://alm.installation.url',
  key: 'github',
  name: 'GitHub'
};

it('should display the provider app install button', () => {
  expect(shallowRender({ boundOrganizations: [] })).toMatchSnapshot();
});

it('should display the bounded organizations dropdown with the list of repositories', () => {
  expect(shallowRender({ organization: 'foo' })).toMatchSnapshot();
});

function shallowRender(props: Partial<AutoProjectCreate['props']> = {}) {
  return shallow(
    <AutoProjectCreate
      almApplication={almApplication}
      boundOrganizations={[
        { alm: { key: 'github', url: '' }, key: 'foo', name: 'Foo' },
        { alm: { key: 'github', url: '' }, key: 'bar', name: 'Bar' }
      ]}
      onProjectCreate={jest.fn()}
      organization=""
      {...props}
    />
  );
}
