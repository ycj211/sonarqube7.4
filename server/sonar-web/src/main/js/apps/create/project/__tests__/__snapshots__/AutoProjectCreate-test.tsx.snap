// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the bounded organizations dropdown with the list of repositories 1`] = `
<Fragment>
  <withRouter(OrganizationInput)
    autoImport={true}
    onChange={[Function]}
    organization="foo"
    organizations={
      Array [
        Object {
          "alm": Object {
            "key": "github",
            "url": "",
          },
          "key": "foo",
          "name": "Foo",
        },
        Object {
          "alm": Object {
            "key": "github",
            "url": "",
          },
          "key": "bar",
          "name": "Bar",
        },
      ]
    }
  />
  <RemoteRepositories
    almApplication={
      Object {
        "backgroundColor": "blue",
        "iconPath": "icon/path",
        "installationUrl": "https://alm.installation.url",
        "key": "github",
        "name": "GitHub",
      }
    }
    onProjectCreate={[MockFunction]}
    organization="foo"
  />
</Fragment>
`;

exports[`should display the provider app install button 1`] = `
<Fragment>
  <p
    className="spacer-bottom"
  >
    onboarding.create_project.install_app_description.github
  </p>
  <IdentityProviderLink
    className="display-inline-block"
    identityProvider={
      Object {
        "backgroundColor": "blue",
        "iconPath": "icon/path",
        "installationUrl": "https://alm.installation.url",
        "key": "github",
        "name": "GitHub",
      }
    }
    onClick={[Function]}
    small={true}
    url="https://alm.installation.url"
  >
    onboarding.import_organization.choose_organization_button.github
  </IdentityProviderLink>
</Fragment>
`;
